dev/tools/lib/repository.dart

```diff
@@ -0,0 +1,357 @@
+// Copyright 2014 The Flutter Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+import 'dart:convert' show jsonDecode;
+import 'dart:io' as io;
+
+import 'package:file/file.dart';
+import 'package:meta/meta.dart';
+import 'package:process/process.dart';
+import 'package:platform/platform.dart';
+
+import './git.dart';
+import './globals.dart' as globals;
+import './stdio.dart';
+import './version.dart';
+
+/// A source code repository.
+class Repository {
+  Repository({
+    @required this.name,
+    @required this.upstream,
+    @required this.processManager,
+    @required this.stdio,
+    @required this.platform,
+    @required this.fileSystem,
+    @required this.parentDirectory,
+    this.localUpstream = false,
+    this.useExistingCheckout = false,
+  })  : git = Git(processManager),
+        assert(localUpstream != null),
+        assert(useExistingCheckout != null);
+
+  final String name;
+  final String upstream;
+  final Git git;
+  final ProcessManager processManager;
+  final Stdio stdio;
+  final Platform platform;
+  final FileSystem fileSystem;
+  final Directory parentDirectory;
+  final bool useExistingCheckout;
+
+  /// If the repository will be used as an upstream for a test repo.
+  final bool localUpstream;
+
+  Directory _checkoutDirectory;
+
+  /// Lazily-loaded directory for the repository checkout.
+  ///
+  /// Cloning a repository is time-consuming, thus the actual repository is not
```

Comment:
ensured to exist -> cloned?

Reply:
Technically, if you pass `useExistingCheckout = true`, it will not necessarily clone every time. Later, when the conductor runs in interactive mode, subsequent commands will operate on the same local repo.

That having been said, I agree "ensured to exist" is confusing. I'll think of a better way to word it.

Reply:
"cloned if it does not already exist"?